cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/utils.go

```diff
@@ -147,23 +122,70 @@ func newPackageRepository(unstructuredRepo map[string]interface{}) (*v1alpha1.Pa
 	if err != nil || !found {
 		return nil, status.Errorf(
 			codes.Internal,
-			"required field metadata.name not found on HelmRepository: %v:\n%v", err, unstructuredRepo)
+			"required field metadata.name not found on HelmRepository: %v:\n%s", err, prettyPrintMap(unstructuredRepo))
 	}
 	namespace, found, err := unstructured.NestedString(unstructuredRepo, "metadata", "namespace")
 	if err != nil || !found {
 		return nil, status.Errorf(
 			codes.Internal,
-			"field metadata.namespace not found on HelmRepository: %v:\n%v", err, unstructuredRepo)
+			"field metadata.namespace not found on HelmRepository: %v:\n%s", err, prettyPrintMap(unstructuredRepo))
 	}
 	url, found, err := unstructured.NestedString(unstructuredRepo, "spec", "url")
 	if err != nil || !found {
 		return nil, status.Errorf(
 			codes.Internal,
-			"required field spec.url not found on HelmRepository: %v:\n%v", err, unstructuredRepo)
+			"required field spec.url not found on HelmRepository: %v:\n%s", err, prettyPrintMap(unstructuredRepo))
 	}
 	return &v1alpha1.PackageRepository{
 		Name:      name,
 		Namespace: namespace,
 		Url:       url,
 	}, nil
 }
+
+// implements plug-in specific cache-related functionality
+// onAddOrModifyRepo essentially tells the cache what to store for a given key
+func onAddOrModifyRepo(key string, unstructuredRepo map[string]interface{}) (interface{}, bool, error) {
+	ready, err := isRepoReady(unstructuredRepo)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if ready {
+		packages, err := indexOneRepo(unstructuredRepo)
+		if err != nil {
+			return nil, false, err
+		}
+		protoMsg := corev1.GetAvailablePackageSummariesResponse{
+			AvailablePackagesSummaries: packages,
+		}
+		bytes, err := proto.Marshal(&protoMsg)
+		if err != nil {
+			return nil, false, err
+		}
+		return bytes, true, nil
+	} else {
+		// repo is not quite ready to be indexed - not really an error condition,
+		// just skip it eventually there will be another event when it is in ready state
+		log.Infof("Skipping packages for repository [%s] because it is not in 'Ready' state", key)
+		return nil, false, nil
+	}
+}
+
+func onGetRepo(key string, value interface{}) (interface{}, error) {
+	bytes, ok := value.([]byte)
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "unexpected value found in cache for key [%s]: %v", key, value)
+	}
+
+	var protoMsg corev1.GetAvailablePackageSummariesResponse
+	err := proto.Unmarshal(bytes, &protoMsg)
+	if err != nil {
+		return nil, err
+	}
+	return protoMsg.AvailablePackagesSummaries, nil
```

Comment:
Hmm, so we're just caching the blob of available package summaries. From a bit of reading, it sounds like we'd need to add our own indicies manually to have some sort of selection/filtering of results with redis, which might be one big reason to use a database for the cache in the first place :/ (just thinking out aloud based on what we're currently doing with the helm support).

Reply:
yeah, I know. I will change what is being cached for flux. Wasn't meant to be final. The goal was to get the plumbing right.